api/v1beta1/openstackdataplanenodeset_types.go

```diff
@@ -177,14 +185,49 @@ func (instance OpenStackDataPlaneNodeSet) GetAnsibleEESpec() AnsibleEESpec {
+	Frr                  string `json:"frr,omitempty"`
+	IscsiD               string `json:"iscsiD,omitempty"`
+	Logrotate            string `json:"logrotate,omitempty"`
+	NeutronMetadataAgent string `json:"neutronMetadataAgent,omitempty"`
+	NovaCompute          string `json:"novaCompute,omitempty"`
+	NovaLibvirt          string `json:"novaLibvirt,omitempty"`
+	OvnControllerAgent   string `json:"ovnControllerAgent,omitempty"`
+	OvnBgpAgent          string `json:"ovnBgpAgent,omitempty"`
+}
+
+const (
+	// FrrDefaultImage -
+	FrrDefaultImage = "quay.io/podified-antelope-centos9/openstack-frr:current-podified"
+	// IscsiDDefaultImage -
+	IscsiDDefaultImage = "quay.io/podified-antelope-centos9/openstack-iscsid:current-podified"
+	// LogrotateDefaultImage -
+	LogrotateDefaultImage = "quay.io/podified-antelope-centos9/openstack-cron:current-podified"
+	// NeutronMetadataAgentDefaultImage -
+	NeutronMetadataAgentDefaultImage = "quay.io/podified-antelope-centos9/openstack-neutron-metadata-agent-ovn:current-podified"
+	// NovaComputeDefaultImage -
+	NovaComputeDefaultImage = "quay.io/podified-antelope-centos9/openstack-nova-compute:current-podified"
+	// NovaLibvirtDefaultImage -
+	NovaLibvirtDefaultImage = "quay.io/podified-antelope-centos9/openstack-nova-libvirt:current-podified"
+	// OvnControllerAgentDefaultImage -
+	OvnControllerAgentDefaultImage = "quay.io/podified-antelope-centos9/openstack-ovn-controller:current-podified"
+	// OvnBgpAgentDefaultImage -
+	OvnBgpAgentDefaultImage = "quay.io/podified-antelope-centos9/openstack-ovn-bgp-agent:current-podified"
+)
+
+// SetupDefaults - initializes any CRD field defaults based on environment variables (the defaulting mechanism itself is implemented via webhooks)
+func SetupDefaults() {
+	// Acquire environmental defaults and initialize dataplane defaults with them
+	dataplaneAnsibleContainerImagesDefaults = DataplaneAnsibleContainerImages{
+		Frr:                  util.GetEnvVar("RELATED_IMAGE_OPENSTACK_EDPM_FRR_DEFAULT_IMG", FrrDefaultImage),
+		IscsiD:               util.GetEnvVar("RELATED_IMAGE_OPENSTACK_EDPM_ISCSID_DEFAULT_IMG", IscsiDDefaultImage),
+		Logrotate:            util.GetEnvVar("RELATED_IMAGE_OPENSTACK_EDPM_LOGROTATE_CROND_DEFAULT_IMG", LogrotateDefaultImage),
+		NeutronMetadataAgent: util.GetEnvVar("RELATED_IMAGE_OPENSTACK_EDPM_NEUTRON_METADATA_AGENT_DEFAULT_IMG", NeutronMetadataAgentDefaultImage),
+		NovaCompute:          util.GetEnvVar("RELATED_IMAGE_OPENSTACK_EDPM_NOVA_COMPUTE_DEFAULT_IMG", NovaComputeDefaultImage),
+		NovaLibvirt:          util.GetEnvVar("RELATED_IMAGE_OPENSTACK_EDPM_LIBVIRT_DEFAULT_IMG", NovaLibvirtDefaultImage),
+		OvnControllerAgent:   util.GetEnvVar("RELATED_IMAGE_OPENSTACK_EDPM_OVN_CONTROLLER_AGENT_DEFAULT_IMG", OvnControllerAgentDefaultImage),
+		OvnBgpAgent:          util.GetEnvVar("RELATED_IMAGE_OPENSTACK_EDPM_OVN_BGP_AGENT_IMAGE", OvnBgpAgentDefaultImage),
+	}
+	SetupDataplaneAnsibleContainerImagesDefaults(dataplaneAnsibleContainerImagesDefaults)
```

Comment:
perhaps so. I believe the idea is that it's needed to set the variable in the webhook

Reply:
Because it's Heap allocated, and the webhook and this file share the same package name. The webhook already has access to the Heap.

To test this theory, I commented it out and then added that unit test I suggested, this function call didn't have any impact on the results of the CR after the webhook had processed it.

But in saying that, I only used the functional test to validate my theory. I didn't actually run the operator and create a nodeset. So I could be overlooking something.